### PR TITLE
ISSUE-27475 - Prevent unnecessary product loads during rule validation

### LINKED ISSUE #27475: Products loaded unnecessarily during rule validation
When any integration uses the Magento rule system (even core sales rules), products may be loaded unnecessarily during rule validation.

Reference:  `\Magento\Rule\Model\Condition\AbstractCondition::validate`

### Preconditions (*)

1. Any Magento OS/EE version >= 2.1 ([reference](https://github.com/magento/magento2/blob/2.1.0/app/code/Magento/Rule/Model/Condition/AbstractCondition.php#L849)), maybe even earlier
2. Any active sales rule using product-attribute based conditions
3. Any attribute of any type supporting and set to _Values Required = No_

### Steps to reproduce (*)

1. Configure a sales rule according to the preconditions
2. For example, create a new attribute, `test_attribute`, type text, values required: no
3. Make sure to configure for use in promo rule conditions
4. Save the attribute (may require assignment to attribute set)
5. Create a new, active sales rule with coupon code (for ease of testing)
6. Configure attribute conditions as shown below and save the rule
7. Ensure sales rule index has been updated (this is assumed)

image

1. Go to the storefront
2. Add any item to the cart
3. At the cart, enter the coupon code (even a wrong code will work)
4. Probe method `\Magento\Rule\Model\Condition\AbstractCondition::validate`
5. Observe check for presence of `test_attribute` key on product object
6. Observe product is loaded from the DB

This illustrates the problem. To prove it another way:

1. Remove the item from your cart
2. In the admin panel, edit the same product
3. Set any value for Test Attribute (does not have to be "X")
4. Save the product
5. Verify any necessary index operations are completed
6. Return to the storefront
7. Add the same item to the cart
8. Again probe method `\Magento\Rule\Model\Condition\AbstractCondition::validate`
9. Observe product is _not_ loaded at all

### Expected result (*)

1. The product model is not loaded during rule validation when it holds no value for the target attribute of the current condition
2. The product model initialized and assigned to the quote item should contain all information necessary to complete rule validation

### Actual result (*)

1. When using attribute-based conditions, the product may be loaded during rule validation

### Additional Information

I discovered this issue in a vendor integration (Amasty Shipping Restrictions) when reviewing performance profiles. Curiously, Magento 1 does not behave this way (see `\Mage_Rule_Model_Condition_Abstract::validate`). I'm not a historian but given the refactor for M2, I'd suggest the developer was either actively solving for a known issue or else attempted a QoL improvement when translating the code.

I don't personally believe that rule validation should have any part in loading models. The requestor should be responsible for supplying all needed information to the validator. And if it isn't present, it should be treated as "not available," but **not attempt** to go load the value. I would even take that so far as to say the `validate` method should expect a more generic `DataObject` or plain DTO as a way to more strictly enforce this idea.

As an extra technical detail, the root cause for this situation is the attribute persister mechanism: `\Magento\Eav\Model\ResourceModel\UpdateHandler::execute` This acts as a self-cleaning mechanism to delete EAV rows that hold "empty-ish" values when saving products. While I agree with this approach, it requires the developer to be aware of its effects (which previously I was _not_!). For example, I would expect that every attribute assigned to a set has a property initialized (even if empty) on product models that are members of the attribute set. Therefore, `DataObject::hasData` should always return true for my attribute check; ie `DataObject::hasData('test_attribute')`. Or else, we redefine the meaning of "has data" (but that's another story).

My point is, the logic in `\Magento\Rule\Model\Condition\AbstractCondition::validate` seems misguided. I believe that a blind check for the value of the attribute here is sufficient. We don't need to know whether the property has been initialized. Let the mixed return type of `getData` and loose type comparison suffice for completing validation. That would be in agreement with the broad usability of the abstract condition validator.

Until this is addressed, large quotes, large catalogs, and shops with many active rules using attribute-based conditions are subject to major performance degradation at scale. And I'll show you what I mean by leaving a screenshot of one transaction trace captured in New Relic.

image

I am showing a 3rd party extension here but please don't close the ticket on account of it! Study the trace and you'll see it leads into the core. And I was also able to reproduce the problem without this vendor integration.

### PR DESCRIPTION
**Pull request by zaximus84**
### Description (*)
When Magento\Rule\Model\Condition\AbstractCondition::validate attempts to validate an entity during rule processing, it does a full load of the entity if the attribute being evaluated is not present in the entity's data array (checked with hasData). For shopping cart price rules, this results in a product being fully loaded multiple times if it doesn't have a value for that attribute in the database.

Magento\SalesRule\Model\Plugin\QuoteConfigProductAttributes ensures that all attributes which might be needed by shopping cart price rules are loaded on the cart products, so in theory, a hasData result of false always means the product doesn't have data for that attribute. $model->load($model->getId()) won't help. The model needs a record of the original data load attempt, even if there was no data for a given attribute.

This PR enhances the attribute value load process in the eav collection abstract to ensure that selected attributes are present in the post-load data array, even if there was no value in the database (in which case the value is set as null).

As an example, let's say we just created the attribute "some_attribute" and it has no default value. No products have been edited to assign a value yet. This is what happens when a product collection is loaded:

```
$productCollection->addAttributeToSelect('some_attribute')->load();
$productCollection->getFirstItem()->hasData('some_attribute'); //false
$productCollection->getFirstItem()->getData(); //null
```

This is what will happen with the changes in this PR:

```
$productCollection->addAttributeToSelect('some_attribute')->load();
$productCollection->getFirstItem()->hasData('some_attribute'); //true
$productCollection->getFirstItem()->getData(); //null
```

Because "some_attribute" is a key in the data array, it's clear that the attribute value was loaded. In Magento\Rule\Model\Condition\AbstractCondition::validate, hasData will be true, and the product will not be unnecessarily loaded again.

### Fixed Issues (if relevant)
1. Fixes magento/magento2#27475

### Manual testing scenarios (*)
See preconditions, steps to reproduce, and expected result in [https://github.com/magento/magento2/issues/27475](https://github.com/magento/magento2/issues/27475)

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Eav/Model/Entity/Collection/AbstractCollection.php [hunk 1]
```diff
@@ -1210,6 +1210,16 @@ public function _loadAttributes($printQuery = false, $logQuery = false)
             }
         }
 
+        foreach ($this->_itemsById as $items) {
```
Comment (swnsma) [inline app/code/Magento/Eav/Model/Entity/Collection/AbstractCollection.php:1214]: I hope we could simplify the code a little but
```
$selectAttributesCodes = array_keys($this->_selectAttributes);
foreach ($this->_itemsById as $items) {
    foreach ($items as $item) {
        foreach (array_diff($selectAttributesCodes, array_keys($item->getData())) as $attributeCode) {
            $item->setData($attributeCode, null);
        }
    }
}
```

**Review by swnsma — CHANGES_REQUESTED** [review]
Hi @zaximus84!

Thank you for your contribution!
I appreciate your PR and think that this is a great thing that is going to deliver.

However, there are a few things I need to ask you:
1) Please check my code suggestion. I believe that this is one of the performance vulnerable places and we need to fight with each function call.
2) I do not like the idea that we will have only this approach for collections. Could you check if we can add the same behavior (set null for missed attributes) also during entities load (I can suggest starting from \Magento\Eav\Model\ResourceModel\ReadHandler::execute)?
3) Also, Magento Definition of Done requires to cover each PR with some kind of autotests. Could you cover new changes with tests?
I believe, adjust certain scenario, which is going to be fixed in current PR - should be enough (\Magento\Eav\Test\Unit\Model\Entity\Collection\AbstractCollectionTest)

**Comment by swnsma** [comment]
@magento run all tests

**Comment by swnsma** [comment]
@magento run all tests

**Comment by swnsma** [comment]
@magento run all tests

**Comment by swnsma** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
Hi @zaximus84, could you please look through failed tests?

**Comment by engcom-Charlie** [comment]
Hi @zaximus84, I'm closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for your collaboration.

**Pull request closed without merging**